// @souche-ui/srn-ui/src/components/List/styles/index.style.js
import { StyleSheet } from 'react-native';
import FontSize from '../../../const/FontSize';
import theme from '../../../theme';
const styles = {
    listView: {
        marginBottom: 16,
    },
    listTitleSpan: {
        height: 16,
    },
    listTitle: {
        marginBottom: 8,
        marginLeft: 16,
        fontSize: FontSize.P2,
        color: theme('color_text_caption'),
    },
    itemViewContainer:{
        backgroundColor: 'white',
    },
    itemView:{
        paddingRight: 16,
        marginLeft: 16,
        borderTopWidth: StyleSheet.hairlineWidth,
        borderTopColor: theme('color_line_b'),
    },
    container:{
        flexDirection: 'row',
        alignItems: 'center',
        height: 48,
    },
    highContainer:{
        flexDirection: 'row',
        justifyContent: 'space-between',
        alignItems: 'center',
        height: 72,
    },
    childItemContainer: {
        minHeight: 48,
        borderTopWidth: StyleSheet.hairlineWidth,
        borderTopColor: theme('color_line_b'),
        marginLeft: 16
    },
    left:{
        flexDirection: 'row',
        alignItems:'center'
    },
    right:{
        flex: 1,
        flexDirection: 'row',
        alignItems:'center'
    },
    rightElement:{
        flex: 1,
        flexDirection: 'row',
        alignItems:'center'
    },
    title:{
        fontSize: FontSize.P1,
        color: theme('color_text_title'),
    },
    subtitle:{
        paddingTop: 10,
        fontSize: FontSize.P2,
        color: theme('color_text_caption'),
    },
    itemIcon: {
        marginRight: 12,
    },
    extraColor: {
        color: theme('color_text_body'),
    },
    disabled: {
        color: theme('color_text_caption'),
    },
    arrow:{
        width: 16,
        height: 16,
    },
    firstItem: {
        borderTopWidth: 0,
        borderTopColor: 'transparent'
    },
    flexGhost: {
        flex: 1,
    },
};


export default styles;
