// @souche-ui/srn-ui/src/components/Switch/Switch.js
import React from 'react';
import {
    Switch as RNSwitch,
    Platform,
} from 'react-native';
import PropTypes from 'prop-types';
import theme from '../../theme';

const Switch = (props)=>{
    let {
        disabled = false,
        checked = false,
        onChange,
        onTintColor,
        thumbTintColor,
        tintColor,
        style,
    } = props;

    if (Platform.OS === 'android' && disabled) {
        onTintColor = theme('color_primary_disable');
        thumbTintColor = null;
        tintColor = null;
    }

    return (
        <RNSwitch
            disabled={disabled}
            onTintColor={onTintColor}
            onValueChange={onChange}
            thumbTintColor={thumbTintColor}
            tintColor={tintColor}
            value={checked}
            style={style}
        />
    );
};

Switch.propTypes = {
    disabled: PropTypes.bool,
    checked: PropTypes.bool,
    onChange: PropTypes.func,
    onTintColor: PropTypes.string,
    thumbTintColor: PropTypes.string,
    tintColor: PropTypes.string,
    style: RNSwitch.propTypes.style,
};

Switch.defaultProps = Platform.select({
    ios: {
        onTintColor: theme('color_primary'),
    },
    android: {
        tintColor: theme('color_background_page'),
        thumbTintColor: theme('color_white'),
        onTintColor: theme('color_primary'),
    },
});

export default Switch;
