// @souche-f2e/srn-navigator/src/TabBar/Badge.js
/**
 * <plusmancn@gmail.com> created at 2017
 *
 * Copyright (c) 2017 Souche.com, all rights
 * reserved.
 *
 * 徽章计数组件
 *
 * @flow
 */
import React, { Component } from 'react';
import {
    StyleSheet,
    Text,
    View
} from 'react-native'; import PropTypes from 'prop-types';

export default class Badge extends Component {
    // width + right = 35，保证左边起始点一致，最多支持 99+
    _computeWidthAndRight(number: number) {
        if (!number) {
            return {
                opacity: 0
            };
        }

        const anchorWidth = 35;
        let right = 0;

        if (number < 10) {
            right = 15;
        } else if (number < 100) {
            right = 8;
        } else {
            right = 0;
        }

        return {
            width: anchorWidth - right,
            right: right
        };
    }

    _renderText(number) {
        if (number > 99) {
            return '99+';
        }

        return number || null;
    }

    render() {
        let { unReadMessageCount } = this.props;
        return (
            <View style={[
                styles.badge,
                this.props.style,
                this._computeWidthAndRight(unReadMessageCount)
            ]}>
                <Text
                    style={[
                        styles.badgeText,
                    ]}
                >
                {this._renderText(unReadMessageCount)}
                </Text>
            </View>
        );
    }
}


const styles = StyleSheet.create({
    badge: {
        alignItems: 'center',
        justifyContent: 'center',
        backgroundColor: '#ff4040',
        overflow: 'hidden',
        height: 19,
        width: 35,
        borderRadius: 10,
        borderWidth: 1,
        borderColor: '#ffffff',
    },
    badgeText: {
        fontSize: 12,
        color: '#ffffff',
    },
});
