// @souche-f2e/srn-navigator/src/TabBar/Tab.js
/**
 * <plusmancn@gmail.com> created at 2017
 *
 * Copyright (c) 2017 Souche.com, all rights
 * reserved.
 *
 * Tab 点击元素
 *
 * @flow
 */

import React, { Component } from 'react';
import {
    TouchableOpacity,
    StyleSheet,
    View,
    Image,
    Text
} from 'react-native';

import Color from '../Color';


import Badge from './Badge.js';
import {NativeModules} from 'react-native';

const AppData = NativeModules.SCCRNAppData;
const app_name = AppData && AppData.appName ? AppData.appName.toLowerCase():'';
export default class Tab extends Component {
    render() {
        let {
            title,
            color,
            tintColor,
            icon,
            tintIcon,
            onPress,
            selected,
            badge,
            textColor
        } = this.props;

        tintColor = tintColor || Color.Primary;
        textColor = tintColor || Color.PrimarySecond;

        color = color || Color.B1;
        let textColorStyle = {
            color: selected ? textColor : color
        };

        return (
            <TouchableOpacity
                activeOpacity={1}
                onPress={this._onPress}
            >
                <View
                    style={styles.container}
                >
                    <Image
                        source={{
                            uri: selected ? tintIcon : icon
                        }}
                        style={[styles.icon]}
                    />
                    <Text
                        style={[
                            styles.text,
                            textColorStyle
                        ]}
                    >
                        {title}
                    </Text>

                    <Badge
                        unReadMessageCount={this.props.badge}
                        style={styles.badge}
                    />
                </View>
            </TouchableOpacity>
        );
    }

    _onPress= (event) => {
        this.props.onPress(event);
    }
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center',
        width: 75
    },
    icon: {
        width: 24,
        height: 24
    },
    text: {
        marginTop: 4,
        fontSize: 10
    },
    badge: {
        position: 'absolute',
        top:  2
    }
});
