// @souche-f2e/srn-navigator/src/TabBar/TabBarItem.js
/**
 * <plusmancn@gmail.com> created at 2017
 *
 * Copyright (c) 2017 Souche.com, all rights
 * reserved.
 *
 * TabBarItem
 */
import {NavigationViewBuilder,Navigator} from '@souche-f2e/srn-navigator';
import React, {
    Component
} from 'react';

import PropTypes from 'prop-types';

export default class TabBarItem extends Component {
    static propTypes = {
        title: PropTypes.string.isRequired,
        color: PropTypes.string,
        tintColor: PropTypes.string,
        icon: PropTypes.string.isRequired,
        tintIcon: PropTypes.string.isRequired,
        selected: PropTypes.bool,
        badge: PropTypes.number,
        onPress: PropTypes.func
    }

    static defaultProps = {
        selected: false
    }
    constructor(props) {
        super(props);
        this.nav = new Navigator(NavigationViewBuilder,{
            backAndroidBind: false
        })
        this.nav.push(this.props.component, this.props.componentProps||{})
    }
    render() {
        return (
          <this.nav.View/>
        )
    }
}
