// @souche-f2e/srn-framework/examples/basic/Page2.js
import React from 'react';

import {
    StyleSheet,
    Text,
    ScrollView,
    TouchableOpacity,
    View
} from 'react-native';
import {TabBar} from '@souche-f2e/srn-navigator';
import {SRNPage} from './../../src/index.js';
import LifeCircle from '../../src/modules/decorators/LifeCircle';
import Tab1 from './tab1';
import Tab2 from './tab2';
import Tab3 from './tab3';
import SRNNative from '@souche-f2e/srn-native'
@LifeCircle
class Page1 extends SRNPage {


    constructor(){

        super();
        this.state = {
            a: 123
        }
    }

    componentDidMount(){
        console.log(this.route)
        this.emitter.on('share', data => {
            this.setState({
                a: 234
            })
        })
        setTimeout(()=>{
            this.setNavigation({
                title:{
                    text:'变化title'
                }
            })
        },1000)
    }

    render(){
        return (<View>
        </View>)
    }
    viewDidAppear() {
        console.log('2 appear')
    }
    viewDidDisappear() {
        console.log('2 disappear')
    }
}

export default Page1;
