// @souche-f2e/srn-framework/examples/basic/Page1.js
import React from 'react';
import NavHelper from '@souche-f2e/srn-navigator';
import {
    StyleSheet,
    Text,
    ScrollView,
    TouchableOpacity,
    View,
    Image
} from 'react-native';

import {SRNPage} from './../../src/index.js';
import IndexStore from '../stores/IndexStore';
import LifeCircle from '../../src/modules/decorators/LifeCircle';

@LifeCircle
class Page1 extends SRNPage {

    static navigation = {
        title: '哈哈哈',
        right: {
            component: ()=>{
                return <Text>aaa</Text>
            },
            onPress: function(emitter, sceneProps) {
                emitter.emit('share', {command: '分享'})
            }
        }
    }

    constructor(){
        super();
        this.state = {
            a: 12,
            images: [

            ]
        }
        this.store = new IndexStore();

    }

    componentDidMount(){
        this.store.fetch();
    }

    viewDidAppear() {
        console.log('1 appear')
    }
    viewDidDisappear() {
        console.log('1 disappear')
    }
    render(){
        return (<View>
            <Text>{this.state.a}</Text>
            <TouchableOpacity onPress={()=>{
                {/*NavHelper.pop()*/}
                this.route.open('/page2',{
                    aa:'value'
                });
            }}>
                <Text>点击跳转第二页</Text>
            </TouchableOpacity>
            <Image source={{uri:'http://img.souche.com/model-material/202719/tgc_dsc_125/20170922152223/7e58b075c8dbe21c0b6e9fc8aac18a23.jpg'}} />
        </View>)
    }
}

export default Page1;
