// @souche-ui/srn-ui/src/components/Radio/Radio.Tag.js
import React from 'react';
import {
    View,
} from 'react-native';
import PropTypes from 'prop-types';

import styleUtil from '../../util/styleUtil';
import Radioable from './Radioable';
import Tag from '../Tag';

function RadioTag(props) {
    const {options, disabled, renderLayout} = props;
    const _renderLayout = renderLayout || renderDefaultLayout;

    return (
        // 同 Radio.Group
        <Radioable {...props}>
            {({value, onChange}) => {
                return _renderLayout(options.map((option, i) => {
                    return (
                        <Tag
                            {...option}
                            key={i}
                            selected={value === option.value}
                            disabled={disabled || option.disabled}
                            onSelect={selected => onChange(selected, option)}
                        >{option.label}</Tag>
                    );
                }));
            }}
        </Radioable>
    );
}

function renderDefaultLayout(tags) {
    return (
        <View style={styles.container}>
            {tags.map((tag, i) => {
                return <View key={i} style={[styles.option]}>
                    {tag}
                </View>;
            })}
        </View>
    );
}

RadioTag.propTypes = {
    value: PropTypes.any,
    defaultValue: PropTypes.any,
    disabled: PropTypes.bool,
    options: PropTypes.array,
    renderLayout: PropTypes.func,
    onChange: PropTypes.func,
};

RadioTag.defaultProps = {
    disabled: false,
    options: [],
};

const styles = styleUtil.create({
    common: {
        container: {
            flexDirection: 'row',
            flexWrap: 'wrap',
            marginTop: 10,
            paddingHorizontal: 16,
            alignItems: 'flex-start',
        },
        option: {
            marginRight: 10,
            marginBottom: 10,
        },
    }
});

export default RadioTag;
