// @souche-ui/srn-ui/src/components/Radio/Radio.Group.js
import React from 'react';
import {
    View,
} from 'react-native';
import PropTypes from 'prop-types';

import styleUtil from '../../util/styleUtil';
import Radioable from './Radioable';
import Radio from './Radio';

function RadioGroup(props) {
    const {options, disabled, span} = props;

    return (
        // 这里对 Radioable 实例化
        // radio 算法是 Radioable 内部实现的，这里只关心 UI 的实现
        <Radioable {...props}>
            {/* 这个函数会在 Radioable 中被调用 */}
            {/* @param value 计算好的当前值 */}
            {/* @param onChange 在实现 UI 时，需要有某个事件被触发并调用此参数，并传入是否选中和选项信息 */}
            {({value, onChange}) => {
                return (
                    <View style={styles.container}>
                        {options.map((option, i) => {
                            return <View key={i} style={[styles.option, { marginRight: span }]}>
                                <Radio
                                    selected={value === option.value}
                                    disabled={disabled || option.disabled}
                                    onChange={selected => onChange(selected, option)}
                                >{option.label}</Radio>
                            </View>;
                        })}
                    </View>
                );
            }}
        </Radioable>
    );
}

RadioGroup.propTypes = {
    value: PropTypes.any,
    defaultValue: PropTypes.any,
    disabled: PropTypes.bool,
    options: PropTypes.array,
    span: PropTypes.number,
    onChange: PropTypes.func,
};

RadioGroup.defaultProps = {
    options: [],
    disabled: false,
    span: 16,
};

const styles = styleUtil.create({
    common: {
        container: {
            flexDirection: 'row',
            alignItems: 'center',
            flexWrap: 'wrap',
            marginVertical: 4,
        },
        option: {
            marginVertical: 8,
        },
    }
});

export default RadioGroup;
