// @souche-f2e/srn-framework/src/modules/SRNPage.js
import React from 'react';
import NavHelper from '@souche-f2e/srn-navigator';
import SRNNative from '@souche-f2e/srn-native';

class SRNPage extends React.Component {
    navigation = {
        title: {},
        left: {},
        right: {}
    }
    route = {
        /**
         * 当前页面的路径
         */
        path: '',
        /**
         * 前一个页面的路径
         */
        referrer: '',
        /**
         * 页面参数
         */
        params: '',
        reload: () => {

        },
        /**
         * 打开某个页面，可以开 RN ，或者 原生 H5（通过协议）
         * @param path
         */
        open: (path, params, callback) => {
            /**
             * 如果是要打开原生协议
             */
            if(typeof(path) === 'string' && /[a-zA-Z]{2,10}\:/.test(path)) {
                if(/^http:|https:/.test(path)) {
                    SRNNative.openWebview(path, params).then(res => {
                        callback(null, res);
                    }, err => {
                        callback(err);
                    });
                } else {
                    SRNNative.open(path, params).then(res => {
                        callback(null, res);
                    }, err => {
                        callback(err);
                    });
                }

            } else {
                // 打开 RN 内其他页面
                NavHelper.push(path, params, callback);
            }

        },
        close: (data) => {
            NavHelper.pop(data);
        }
    };

    static hook = {
        viewDidLoad: null,
        viewWillAppear: null,
        viewDidAppear: null,
        viewWillDisappear: null,
        viewDidDisappear: null,
        dealloc: null
    }

    navigator = NavHelper;

    constructor (props) {
        super();
    }

    viewDidLoad () {

    }

    viewDidAppear () {

    }

    viewDidDisappear () {

    }

    dealloc () {

    }
}

export default SRNPage;
