// @souche-f2e/srn-framework/examples/stores/IndexStore.js
/**
 * 此处定义页面数据管理层，业务复杂请分文件夹，具体文档：http://fedoc.sqaproxy.souche.com/rn/develop/dev_4.html
 */
import { observable } from 'mobx';
import { SRNStore } from '../../src/index';
// 引入model定义，一个页面应该对应一个store，页面内的数据对象则应该定义为model
import CarModel from '../models/CarModel';


import Axios from "axios";


class IndexStore extends SRNStore {
    recommendCar = new CarModel();

    constructor () {
        super();
    }

    /**
     * 请求数据
     */
    fetch () {
                 // 文档：http://fedoc.sqaproxy.souche.com/rn/develop/dev_data.html
                 this.recommendCar.__map({
                     distance: "1_$m",
                     price: "100_$y",
                     name: "fff",
                     user_title: "李老板"
                 });
                 console.log(JSON.stringify(this.recommendCar));

                //  this.__fetch("http://dic.souche.com/api/v1/queryDictionaries/load.json?type=area")

                 this.__axios.get("http://dic.souche.com/api/v1/queryDictionaries/load.json?type=area").then(res=>{
                     console.log(res)
                 })
                //  拦截器
                // this.__axios.interceptors.request.use(function (config) {
                //     console.log('before',config);
                //     return config;
                // }, function (error) {
                //     return Promise.reject(error);
                // });

             }
}

export default IndexStore;
