// @souche-ui/srn-ui/src/components/HDView/HDView.js
import React from 'react';
import {
    View,
    StyleSheet,
} from 'react-native';

/**
 * 用于解决官方 View 组件边框锯齿问题
 */
function HDView(props) {
    const { type: Comp, style, children, ...restProps } = props;
    const {
        position,
        top,
        right,
        bottom,
        left,
        borderColor,
        borderWidth,
        borderRadius,
        margin,
        marginTop,
        marginRight,
        marginBottom,
        marginLeft,
        flex,
        alignSelf,
    } = StyleSheet.flatten(style);
    const wrapperStyle = {
        position,
        top,
        right,
        bottom,
        left,
        backgroundColor: borderColor,
        padding: borderWidth,
        borderRadius: borderRadius + borderWidth,
        margin,
        marginTop,
        marginRight,
        marginBottom,
        marginLeft,
        flex,
        alignSelf,
    };

    return (
        <View style={[wrapperStyle]}>
            <Comp style={[style, styles.reset]} {...restProps}>{children}</Comp>
        </View>
    );
}

HDView.defaultProps = {
    type: View,
};

const styles = StyleSheet.create({
    reset: {
        borderWidth: 0,
        position: undefined,
        top: 0,
        right: 0,
        bottom: 0,
        left: 0,
        margin: 0
    },
});

export default HDView;
