// @souche-f2e/srn-toolchain/lib/generate_module_info.js
'use strict';

const fs = require('fs');
const path = require('path');

const gitUtil = require('./git_util.js');
const globalData = require('./global_data.js');

const pkg = require(path.resolve(globalData.PROJECT_PACKAGE_JSON));

const updateNativeRouter = require("./update_router").updateNativeRouter;


module.exports = {
    run: function (options, done) {
        const branch = gitUtil.getBranch();
        const hosts = this.getHostConfig();

        const moduleInfo = {
            name: pkg.name,
            branch: branch,
            version: options.version,
            hosts: hosts
        };

        try {
            const fileContent = JSON.stringify(moduleInfo, null, 4);
            fs.writeFileSync(globalData.MODULE_INFO_PATH, fileContent);
            updateNativeRouter();
            return done();
        } catch (e) {
            return done(e);
        }
    },
    generateModuleInfoForStart () {
        let branch = '';
        try {
            branch = gitUtil.getBranch();
        } catch (e) {

        }
        const hosts = this.getHostConfig();

        const moduleInfo = {
            name: pkg.name,
            branch: branch,
            version: pkg.version,
            hosts: hosts
        };

        try {
            const fileContent = JSON.stringify(moduleInfo, null, 4);
            fs.writeFileSync(globalData.MODULE_INFO_PATH, fileContent);
        } catch (e) {

        }
    },
    getHostConfig () {
        const HOST_CONFIG_PATH = path.resolve(globalData.PROJECT_ROOT, 'src/config/host.config.js');
        let hostConfig = {};

        try {
            hostConfig = require(HOST_CONFIG_PATH);
        } catch (e) {
        }

        return Object.keys(hostConfig);
    }
};
