// @souche-ui/srn-ui/src/components/AlphList/AlphList.js
import React from 'react';
import {
    View,
    Text,
    TouchableWithoutFeedback,
    PanResponder,
    Platform,
    TextInput,
    StyleSheet,
    Dimensions,
} from 'react-native';
import PropTypes from 'prop-types';

import FontSize from '../../const/FontSize';
import theme from '../../theme';

const WINDOW_WIDTH = Dimensions.get('window').width;

export default class AlphList extends React.Component{
    constructor(props){
        super(props);
        this.state = {
            iconShow: false
        };
        this.selectedIndex = this._getIndexFromValue(this.props.defaultValue, this.props.list);
        this.moving = false;  // 用于标注 panResponder 是否在 moving
    }
    setSelectedValue(value){
        if (!this.moving){
            this._changeSelectedIndex(this._getIndexFromValue(value, this.props.list));
        }
    }
    componentWillMount(){
        this._panResponder = PanResponder.create({
            // 要求成为响应者：
            onStartShouldSetPanResponder: (evt, gestureState) => true,
            onStartShouldSetPanResponderCapture: (evt, gestureState) => false,
            onMoveShouldSetPanResponder: (evt, gestureState) => true,
            onMoveShouldSetPanResponderCapture: (evt, gestureState) => true,

            /*
            * 设置不捕获手势监听，onPressIn 优先执行，设置好当前的 selectedIndex; 但是 pressIn 后, move 之前, onPanResponderGrant 不会触发
            * 在手势开始回调中记录当前 selectedIndex 到 this._gestureStartIndex
            * 在手势滑动时通过滑动距离和 this._gestureStartIndex 计算实时选中值
            * */
            onPanResponderGrant: (evt, gestureState) => {
                // 开始手势操作。
                this._gestureStartIndex = this.selectedIndex;
                if (!this.state.iconShow){
                    this.setState({iconShow: true});
                }
                this.moving = true;
            },
            /*
            * 手指在索引栏移动时通过垂直方向上的移动距离来计算当前选中值
            * 在 ios 中，evt.nativeEvent.locationY 是触点距离 刚开始移动的那一刻的触点的 textContainer 的距离
            * 在 android 中，evt.nativeEvent.locationY 一直是首次触点距离其 textContainer 的距离，不会随着滑动变化
            * gestureState.dy 为滑动距离
            * */
            onPanResponderMove: (evt, gestureState) => {
                // 移动手指，选中索引变化
                let list = this.props.list;
                let index;
                if (Platform.OS === 'ios') {
                    index = this._getIndexFromY(evt.nativeEvent.locationY) + this._gestureStartIndex;
                } else {
                    index = this._getIndexFromY(gestureState.dy) + this._gestureStartIndex;
                }
                if (index < 0) {index = 0;}
                if (index > list.length - 1) {index = list.length - 1;}
                if (index !== this.selectedIndex) {
                    this._changeSelectedIndex(index);
                    this.props.onValueChange && this.props.onValueChange(this.props.list[index], index);
                }
            },
            onPanResponderTerminationRequest: (evt, gestureState) => true,
            onPanResponderRelease: (evt, gestureState) => {
                // 手指离开屏幕
                if (this.state.iconShow){
                    this.setState({iconShow: false});
                }
                this.moving = false;
            },
        });
    }
    render(){
        return (
            <View style={styles.alphlistContainer}>
                <View>
                    <View {...this._panResponder.panHandlers}>
                        {this.props.list.map((item, index)=>{
                            const selectedContainerStyle = this.selectedIndex === index ? {backgroundColor: theme('color_background_page')} : null;
                            return (
                                <TouchableWithoutFeedback key={index} onPressIn={this._onPressIn.bind(this, index)} onPressOut={this._onPressOut.bind(this)} onLongPress={this._onLongPress.bind(this)}>
                                    <View ref={(com)=>{this['_view' + index] = com;}} key={index} style={[styles.textContainer, selectedContainerStyle]}>
                                        <Text style={styles.text}>{item}</Text>
                                    </View>
                                </TouchableWithoutFeedback>
                            );
                        })}
                    </View>
                    {
                        this.state.iconShow ?
                            <View style={[styles.floatIcon, {top: -43 + 18 * (this.props.list.length / 2)}]}>
                                <View style={styles.rectangle}>
                                    <TextInput underlineColorAndroid={'transparent'} ref={(com)=>{this._floatText = com;}} editable={false} style={styles.rectangleText} defaultValue={this.props.list[this.selectedIndex]}/>
                                </View>
                            </View>
                            :
                            null
                    }
                </View>
            </View>
        );
    }
    _getIndexFromY(dy){
        return Math.floor(dy / 18);
    }
    _onPressIn(index){
        this._changeSelectedIndex(index);
        this.props.onValueChange && this.props.onValueChange(this.props.list[index], index);
    }
    _onLongPress(){
        if (!this.state.iconShow){
            this.setState({
                iconShow: true
            });
        }
    }
    _onPressOut(){
        /*
        * 在 move 的过程中会触发 onPressOut ,所以用 this.moving 来标注是否发生 move
        * 在没有发生 move 时释放点击，隐藏 icon
        * 在发生了 move 后，隐藏 icon 交由 onPanResponderRelease
        * */
        if (!this.moving && this.state.iconShow){
            this.setState({
                iconShow: false
            });
        }
    }
    /*
    * @params value: 数组中的一项; list: 数组
    * @return (number): value 在 list 中的索引
    * 如果数组中没有 value 返回 0
    * */
    _getIndexFromValue(value, list){
        let i;
        for (i = 0; i < list.length; i++){
            if (value === list[i]) {
                break;
            }
        }
        if (i >= list.length) {i = 0;}
        return i;
    }
    /**
     * 修改索引的选中项
     * 修改样式，修改 this.selectedIndex, 修改 floatIcon 的 top 值和 text 值
     * */
    _changeSelectedIndex(index){
        let oldSelectedIndex = this.selectedIndex;
        this['_view' + oldSelectedIndex].setNativeProps({
            style: {
                backgroundColor: 'transparent'
            }
        });
        this['_view' + index].setNativeProps({
            style: {
                backgroundColor: theme('color_background_page'),
            }
        });
        if (this.state.iconShow){
            this._floatText.setNativeProps({
                text: this.props.list[index]
            });
        }
        this.selectedIndex = index;
    }
}

AlphList.propTypes = {
    list: PropTypes.array,
    defaultValue: PropTypes.string,
    onValueChange: PropTypes.func,
};

const styles = StyleSheet.create({
    alphlistContainer: {
        alignSelf: 'stretch',
        justifyContent: 'center',
        // backgroundColor: theme('color_white'),    // bug: 当设置背景色后，安卓下 floatIcon 会不显示
        paddingLeft: 2,
        paddingRight: 2,
    },
    textContainer: {
        width: 18,
        height: 18,
        borderRadius: 9,
        justifyContent: 'center',
        alignItems: 'center'
    },
    text: {
        color: theme('color_text_caption'),
        fontSize: FontSize.P3,
    },
    floatIcon: {
        flexDirection: 'row',
        justifyContent: 'center',
        alignItems: 'center',
        position: 'absolute',
        top: -96,
        left: -WINDOW_WIDTH / 2,
    },
    rectangle: {
        width: 86,
        height: 86,
        borderRadius: 4,
        justifyContent: 'center',
        alignItems: 'center',
        backgroundColor: theme('color_text_title'),
    },
    rectangleText: {
        width: 86,
        height: 48,
        fontSize: 32,
        color: theme('color_white'),
        padding: 0,
        textAlign: 'center',
    }
});
