// @souche-ui/srn-ui/src/components/Calendar/MonthHeader.js
import React from 'react';
import {
    View,
    Text,
    StyleSheet,
} from 'react-native';
import FontSize from '../../const/FontSize';
import theme from '../../theme';

const weekNumbers = ['日', '一', '二', '三', '四', '五', '六'];

function MonthHeader() {
    return (
        <View style={styles.container}>
            {weekNumbers.map(n => {
                return (
                    <View key={n} style={styles.weekNumber}>
                        <Text style={styles.weekNumberText}>{n}</Text>
                    </View>
                );
            })}
        </View>
    );
}

const styles = StyleSheet.create({
    container: {
        flexDirection: 'row',
        paddingBottom: 16,
        borderBottomWidth: StyleSheet.hairlineWidth,
        borderBottomColor: theme('color_line_b'),
    },
    weekNumber: {
        flex: 1,
        alignItems: 'center',
        justifyContent: 'center',
    },
    weekNumberText: {
        color: theme('color_text_caption'),
        fontSize: FontSize.P3,
    }
});

export default MonthHeader;
