// @souche-f2e/srn-native/lib/img.js
import { SendModule, State } from '../const';
/** 
 * 图片类 
 */
export default {
    // 上传图片
    uploadImg(params) {
        let data = {
            maxPicCount: 1,
            needEdit: 0
        };
        if(typeof params === 'object') {
            params.needEdit = params.needEdit ? 1 : 0;
            data = Object.assign({}, data, params);
        }
        if(typeof params === 'number') {
            data.maxPicCount = params;
        }
        return SendModule('open/capturePic', data).then(res => {
            return Promise.resolve(res.urls);
        });
    },
    /**
     * @desc 图片裁剪
     * @param {String}   params    图片地址  
     * @param {Number}   width  图片结果的宽
     * @param {Number}   height 图片结果的高
     * @param {Function} cb      回调函数
     *       回调数据： 图片的url
     */
    cutImg(params, width, height) {
        if(typeof params !== 'object') {
            params = {
                width: width,
                height: height,
                imageUrl: params
            }
        }
        return SendModule('open/cutImage', params).then(res => {
            if(res.result === State.SUCCESS) {
                return Promise.resolve(res.url);
            }
            return Promise.reject(res.result);
        });
    },
    /**
     * @desc  查看大图
     * @param {Object}   params    查看大图的资源
     *      pics            {Array}    url路径数组       必传
     *      index           {Number}   当前图片位置      可选   默认： 0
     *      enableDelete    {Boolean}  是否允许操作删除   可选  默认: false
     */
    seeBigImg(data) {
        data.enableDelete = data.enableDelete ? 1 : 0;
        data.index = data.index || 0;
        data.together = 1;
        try {
            data.pics = JSON.parse(JSON.stringify(data.pics));
        } catch(e) {
            data.pics = [];
        }
        return SendModule('open/browserPic', data).then(res => {
            return Promise.resolve(res.urls);
        });
    },

    /**
     * @desc 保存图片到本地
     * @param {Object}   data       参数
     * @param {String}   data.url   图片地址
     */
    saveImage(data) {
        return SendModule('open/saveimage', data);
    },
}
