// @souche-ui/srn-ui/src/components/ActionSheet/ActionSheetItem.js
import React from 'react';
import {
    View,
    Text,
    TouchableHighlight,
    StyleSheet,
} from 'react-native';
import PropTypes from 'prop-types';
import theme from '../../theme';
import FontSize from '../../const/FontSize';

function ActionSheetItem(props) {
    const {
        _index,
        content,
        prefix,
        suffix,
        destructive,
        onPress,
    } = props;

    return (
        <TouchableHighlight
            activeOpacity={1}
            underlayColor={theme('color_press')}
            style={[styles.container, _index === 0 && styles.first]}
            onPress={onPress}
        >
            <View style={styles.contentWrapper}>
                {prefix && React.cloneElement(prefix, { style: [prefix.props.style, styles.prefix] })}
                <Text style={[styles.content, destructive && styles.destructive]}>{content}</Text>
                {suffix && React.cloneElement(suffix, { style: [suffix.props.style, styles.suffix] })}
            </View>
        </TouchableHighlight>
    );
}

ActionSheetItem.propTypes = {
    content: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.element,
    ]),
    prefix: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.element,
    ]),
    suffix: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.element,
    ]),
    destructive: PropTypes.bool,
    onPress: PropTypes.func,
};

ActionSheetItem.defaultProps = {
    destructive: false,
};

const styles = StyleSheet.create({
    container: {
        minHeight: 48,
        backgroundColor: theme('color_white'),
        borderTopWidth: StyleSheet.hairlineWidth,
        borderTopColor: theme('color_line_b'),
        alignItems: 'center',
        justifyContent: 'center',
    },
    first: {
        borderTopWidth: 0,
    },
    content: {
        color: theme('color_text_title'),
        fontSize: FontSize.P1,
    },
    contentWrapper: {
        flexDirection: 'row',
        alignItems: 'center',
        justifyContent: 'center',
    },
    destructive: {
        color: theme('color_warning'),
    },
    prefix: {
        marginRight: 8,
    },
    suffix: {
        marginLeft: 8,
    },
});

export default ActionSheetItem;
