// @souche-f2e/srn-navigator/examples/Demo3.js
import React from 'react';
import Demo4 from './Demo4';
import NavHelper from './../src/index';

import {
    StyleSheet,
    Text,
    Image,
    ScrollView,
    TouchableOpacity
} from 'react-native';

class Demo3 extends React.Component {

    static navigation = {
        title:{
            text: "第1页"
        },
        right: {
            text:"aaa",
            onPress: function(emitter, sceneProps) {
                emitter.emit('share', {command: '分享'})
            }
        }
    }
    constructor(props){
        super(props);
        console.log(props);

    }
    componentWillMount () {
        // console.log(this.route)
        // console.log(this.props)
        this.emitter.on('share', data => {
            console.log(data)
        })
        setTimeout(()=>{
            this.setNavigation({
                title:{
                    text:'变化title'
                }
            })
        },1000)

        this.emitter.on('view_disappear', data => {
            console.log('3 disappear')
        })
        this.emitter.on('view_appear', data => {
            console.log('3 appear')
        })
    }
    render() {
        return (
            <ScrollView style={styles.container}>
                <Text style={styles.text}>第1页</Text>
                <TouchableOpacity onPress={()=>{
                    NavHelper.push(Demo4,{
                        aa:'value'
                    }, function(data){
                        console.log('back from demo2')
                        console.log(data)
                    });
                }}>
                    <Text>点击跳转第4页</Text>
                </TouchableOpacity>

            </ScrollView>

        );
    }
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        padding: 16,
        flexDirection: 'column'
    },
    text: {
        marginBottom: 16,
        lineHeight: 20
    },
});


export default Demo3;
