// @souche-f2e/srn-navigator/examples/Demo.js
import React from 'react';
import Demo2 from './Demo2';
import Demo3 from './Demo3';
import NavHelper from './../src/index';

import {
    StyleSheet,
    Text,
    Image,
    ScrollView,
    TouchableOpacity
} from 'react-native';

class Demo extends React.Component {

    static navigation = {
        title: {
            text: "第1页"
        },
        right: {
            text: "aaa",
            onPress: function (emitter, sceneProps) {
                emitter.emit('share', { command: '分享' })
            }
        },
    }

    constructor (props) {
        super(props);
        console.log(props);

    }

    componentDidMount () {
        // console.log(this.route)
        // console.log(this.props)
        this.emitter.on('share', data => {
            console.log(data)
        })
        // setTimeout(() => {
        //     this.setNavigation({
        //         title: {
        //             text: '变化title'
        //         }
        //     })
        // }, 1000)

        this.emitter.on('view_disappear', data => {
            console.log('1 disappear')
        })
        this.emitter.on('view_appear', data => {
            console.log('1 appear')
        })
    }

    render () {
        return (

          <ScrollView style={styles.container}>
              <Text style={styles.text}>第1页</Text>
              <TouchableOpacity onPress={() => {
                  NavHelper.push(Demo2, {
                      aa: 'value'
                  }, 'first',(data)=>{
                      console.log('callback ',data)
                  });
              }}>
                  <Text style={{fontSize:18}}>点击跳转第2页</Text>

              </TouchableOpacity>
              <TouchableOpacity onPress={() => {
                  NavHelper.push('/demo3', {
                      aa: 'value'
                  }, 'first',(data)=>{
                      console.log('callback ',data)
                  });
              }}>
                  <Text>点击跳转第3页</Text>
              </TouchableOpacity>
              <TouchableOpacity onPress={() => {
                  this.hideNavigation()
              }}>
                  <Text>隐藏 header</Text>
              </TouchableOpacity>
              <TouchableOpacity onPress={() => {
                  this.hideNavigation(true)
              }}>
                  <Text>隐藏 header，保留 statuebar</Text>
              </TouchableOpacity>
              <TouchableOpacity onPress={() => {
                  this.showNavigation()
              }}>
                  <Text>显示 header</Text>
              </TouchableOpacity>
          </ScrollView>

        );
    }
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        padding: 16,
        flexDirection: 'column'
    },
    text: {
        marginBottom: 16,
        lineHeight: 20
    },
});


export default Demo;
