// @souche-ui/srn-ui/src/components/PullToRefresh/PullUpIndicator.js
import React from 'react';
import {
    View,
    Text,
    Animated,
    Easing,
    StyleSheet,
} from 'react-native';
import PropTypes from 'prop-types';
import FontSize from '../../const/FontSize';
import theme from '../../theme';

class PullUpIndicator extends React.PureComponent {

    rotate = new Animated.Value(0);
    shouldAnimationEnd = false;

    componentDidMount() {
        this.startLoadingAnimation();
    }

    componentWillUnmount() {
        this.shouldAnimationEnd = true;
    }

    startLoadingAnimation() {
        this.rotate.setValue(0);
        if (this.shouldAnimationEnd) {
            return;
        }
        Animated.timing(
            this.rotate,
            {
                toValue: 1,
                duration: 1000,
                isInteraction: false,
                easing: Easing.linear,
                useNativeDriver: true,
            }
        ).start(() => { this.startLoadingAnimation(); });
    }

    render() {
        return (
            <View style={styles.indicator}>
                {!this.props.finished ? <Animated.Image
                    style={[styles.indicatorImg, {
                        transform: [{
                            rotate: this.rotate.interpolate({
                                inputRange: [0, 1],
                                outputRange: ['0deg', '360deg']
                            })
                        }]
                    }]}
                    source={{ uri: 'https://assets.souche.com/assets/sccimg/common/loading.png' }}
                    resizeMode={'contain'}
                /> : null}
                <Text style={styles.indicatorText}>{this.props.finished ? '没有更多内容啦~' : '加载中'}</Text>
            </View>
        );
    }
}

PullUpIndicator.propTypes = {
    finished: PropTypes.bool,
};

PullUpIndicator.defaultProps = {
    finished: false,
};

const styles = StyleSheet.create({
    indicator: {
        height: 48,
        backgroundColor: theme('color_background_page'),
        alignItems: 'center',
        justifyContent: 'center',
        flexDirection: 'row'
    },
    indicatorImg: {
        width: 16,
        height: 16,
        tintColor: theme('color_text_caption'),
    },
    indicatorText: {
        color: theme('color_text_caption'),
        fontSize: FontSize.P2,
        marginLeft: 10,
    },
});

export default PullUpIndicator;
