// @souche-f2e/srn-native/__mocks__/index.js
export default {
    // lib/common.js
    relogin() {
        return new Promise((resolve, reject) => {
            this.open('mock://open.present/login');
            resolve({});
        });
    },
    open(protocol, data = {}) {
        return new Promise((resolve, reject) => {
            resolve();
        });
    },
    popAndOpen(protocol, data = {}){
        return Promise.resolve({});
    },
    popAllAndOpen(protocol, data = {}){
        return Promise.resolve({});
    },
    openWebview(url, data = {}) {
        return Promise.resolve({});
    },
    // 新用户引导组件
    openGuideView (data = []) {
        return Promise.resolve({});
    },
    openDatePicker(params = {}) {
        if(typeof params === 'object') {
            params = Object.assign({
                format: "yyyy/MM/dd",
                rightText: '确定',
                leftText: '',
                rightColor: '#FF5050',
                rightSize: '14'
            }, params);
        } else {
            console.err('openDatePicker: 参数异常', params);
        }
        return Promise.resolve('2019/12/26');
    },
    // 获取联系人信息
    getContacts() {
        return Promise.resolve([
            {
                "name": "John Appleseed",
                "phones": "888-555-5512,888-555-1212",
                "section": "J"
            },
            {
                "name": "David Taylor",
                "phones": "555-610-6679",
                "section": "D"
            },
            {
                "name": "Daniel Higgins Jr.",
                "phones": "555-478-7672,(408) 555-5270,(408) 555-3514",
                "section": "D"
            },
            {
                "name": "Hank M. Zakroff",
                "phones": "(555) 766-4823,(707) 555-1854",
                "section": "H"
            },
            {
                "name": "Kate Bell",
                "phones": "(555) 564-8583,(415) 555-3695",
                "section": "K"
            },
            {
                "name": "Anna Haro",
                "phones": "555-522-8243",
                "section": "A"
            }
        ]);
    },
    // 设置联系人信息
    saveContacts(data, flag) {
        if(typeof data !== 'object') {
            console.err('设置联系人数据异常');
            return;
        }
        return Promise.resolve({
            "data": [
                {
                    "phone": "18767136122",
                    "name": "小明",
                    "saved": true
                }
            ]
        });
    },
    // 埋点
    bury(typeId, vals = {}) {
        return Promise.resolve({});
    },
    // 页面埋点
    buryPage(pageID, pageAction, pageType = 3) {
        return Promise.resolve({});
    },
    //Native日志透传
    log(level, msg, tag = "", exception) {
        return Promise.resolve({});
    },
    // 获取渠道用户信息
    getChannelUserInfo(channel = "") {
        return Promise.resolve({});
    },

    // lib/getData.js
    getAppData() {
        return Promise.resolve({
            "loginAccount": "11111111111",
            "shopCode": "01222222",
            "appScheme": "dfc",
            "appVersion": "6.5.6",
            "primaryColor": "#FF571A",
            "env": "2",
            "appName": "dafengche",
            "APP_VIN_SUPPORT": "1",
            "userToken": "01_EVR6_xtCn3aAbR",
            "id": "cyWm4rNnT0",
            "iid": "2222222"
        });
    },
    getGPSData() {
        return Promise.resolve({
            "geopoint": "5.3E-1631,3330.95",
            "state": "浙江省",
            "city": "杭州市"
        });
    },
    getDeviceData() {
        return Promise.resolve({
            "systemVersion": "8.0.0",
            "model": "msm8996ONEPLUS A3000",
            "systemName": "Android OS"
        });
    },

    // lib/img.js
    uploadImg(params) {
        return Promise.resolve([
            'https://picsum.photos/200',
            'https://picsum.photos/300',
        ]);
    },
    /**
     * @desc 图片裁剪
     * @param {String}   params    图片地址  
     * @param {Number}   width  图片结果的宽
     * @param {Number}   height 图片结果的高
     * @param {Function} cb      回调函数
     *       回调数据： 图片的url
     */
    cutImg(params, width, height) {
        return Promise.resolve('https://picsum.photos/200');
    },
    /**
     * @desc  查看大图
     * @param {Object}   params    查看大图的资源
     *      pics            {Array}    url路径数组       必传
     *      index           {Number}   当前图片位置      可选   默认： 0
     *      enableDelete    {Boolean}  是否允许操作删除   可选  默认: false
     */
    seeBigImg(data) {
        return Promise.resolve([
            'https://picsum.photos/200',
            'https://picsum.photos/300',
        ]);
    },
    /**
     * @desc 保存图片到本地
     * @param {Object}   data       参数
     * @param {String}   data.url   图片地址
     */
    saveImage(data) {
        return Promise.resolve({});
    },

    // lib/prompt.js
    toast(text = "", icon = "", duration = 2000, qrcodeText = "") {
        return Promise.resolve({});
    },
    Loading: {
        show(text = "", icon = "") {
            return Promise.resolve({});
        },
        hide() {
            return Promise.resolve({});
        }
    },
    alert(title = "", text = "", button = "确定") {
        return Promise.resolve({});
    },
    confirm(title = "", text = "", leftButton = "取消", rightButton = "确定") {
        return Promise.resolve({});
    },
    setDefaultLoadingIcon(type) {
    },

    // lib/share.js
    share(data) {
        return Promise.resolve(275);
    },

    // lib/other.js
    closeView(data) {
    },
};
