// @souche-ui/srn-ui/src/components/Form/FormView.js
import React from 'react';
import {
    View,
    Text,
    StyleSheet,
} from 'react-native';
import PropTypes from 'prop-types';

import FontSize from '../../const/FontSize';
import theme from '../../theme';
import styleUtil from '../../util/styleUtil';

function FormView(props) {
    const {
        title,
        desc,
        children,
        crowd,
        itemViewStyle,
    } = props;

    return (
        <View style={styles.container}>
            {crowd ? <View style={styles.span} /> : null}
            {/* 表单名称 */}
            {title ? <Text style={styles.title}>{title}</Text> : null}
            {/* 表单主体 */}
            <View style={[styles.panel, itemViewStyle]}>
                {React.Children.toArray(children).filter(c => !!c).map((child, i) => {
                    if (i === 0) {
                        return child;
                    } else {
                        return [<View style={styles.divide}/>, child];
                    }
                })}
            </View>
            {/* 表单提示信息 */}
            {desc ? <Text style={styles.desc}>{desc}</Text> : null}
        </View>
    );
}

FormView.propTypes = {
    title: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.element
    ]),
    desc: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.element
    ]),
    itemViewStyle: View.propTypes.style,
    crowd: PropTypes.bool,
};

FormView.defaultProps = {
    crowd: false,
};

const styles = styleUtil.create({
    common: {
        container: {
            marginBottom: 16,
        },
        holdBelow: {
            height: 16,
        },
        panel: {
            backgroundColor: theme('color_white'),
            minHeight: 48,
        },
        title: {
            color: theme('color_text_caption'),
            fontSize: FontSize.P2,
            marginBottom: 8,
            paddingHorizontal: 16,
        },
        desc: {
            color: theme('color_text_caption'),
            fontSize: FontSize.P2,
            marginTop: 8,
            paddingHorizontal: 16,
        },
        divide: {
            backgroundColor: theme('color_line_b'),
            height: StyleSheet.hairlineWidth,
            marginLeft: 16,
        },
        span: {
            height: 16,
        },
    },
});

export default FormView;
