// @souche-ui/srn-ui/src/components/Popup/Popup.js
import React from 'react';
import {
    View,
    Modal,
    TouchableOpacity,
    Animated,
    StyleSheet,
    Dimensions,
    Platform,
    InteractionManager,
} from 'react-native';
import PropTypes from 'prop-types';
import { isIPhoneX } from '../../util/deviceUtil';
import theme from '../../theme';

const WINDOW_HEIGHT = Dimensions.get('window').height;
/**
 * 动画状态
 */
const ANIMATE_STATUS = {
    /**
     * 空闲
     */
    IDLE: 0,
    /**
     * “显示”动画过程中
     */
    SHOWING: 1,
    /**
     * “隐藏”动画过程中
     */
    HIDDING: 2,
};

class Popup extends React.Component {

    _contentHeight = WINDOW_HEIGHT;
    _layouted = false;
    _onceLayouted = false;
    _animateStatus = ANIMATE_STATUS.IDLE;

    constructor(props) {
        super(props);

        this.state = {
            visible: props.visible,
            contentHeight: WINDOW_HEIGHT + 100,
            popupAnimatedValue: new Animated.Value(0),
        };
        this._backdropAnimatedValue = new Animated.Value(props.visible ? 1 : 0);

        if (this.state.visible) {
            // 如果初始状态 visible 为 true，则暂存“显示”操作，等 layout 完成后执行“显示”操作
            this._showActionStashed = true;
        }

        this._handleLayout = this._handleLayout.bind(this);
        this._handleRequestClose = this._handleRequestClose.bind(this);
    }

    componentWillReceiveProps(nextProps) {
        if (nextProps.visible !== this.props.visible) {
            if (nextProps.visible) {
                this.setState({
                    visible: true,
                });
                this._show();
            } else {
                this._hide(() => {
                    this.setState({
                        visible: false,
                    });
                });
            }
        }
    }

    _adjustPosition() {
        if (this._animateStatus === ANIMATE_STATUS.IDLE) {
            this.state.popupAnimatedValue.setValue(-this._contentHeight);
        }
    }

    _show() {
        if (!this._layouted) {
            // 如果还没有 layout，则暂存“显示”操作，等 layout 完成后执行“显示”操作
            this._hideActionStashed = false;
            this._showActionStashed = true;
            return;
        }
        if (this._animateStatus === ANIMATE_STATUS.SHOWING) {
            // 防止重复动画
            return;
        }
        this._hideAnimation && this._hideAnimation.stop();
        this._animateStatus = ANIMATE_STATUS.SHOWING;
        // 注册动画，方便调度
        this._interactionHandler && InteractionManager.clearInteractionHandle(this._interactionHandler);
        this._interactionHandler = InteractionManager.createInteractionHandle();
        Animated.timing(
            this._backdropAnimatedValue,
            {
                toValue: 1,
                duration: 300,
                useNativeDriver: true,
            }
        ).start();
        this._showAnimation = Animated.timing(
            this.state.popupAnimatedValue,
            {
                toValue: -this._contentHeight,
                duration: 300,
                useNativeDriver: Platform.OS !== 'ios' ? true : false,
            }
        );
        this._showAnimation.start(() => {
            // 动画结束
            this._animateStatus = ANIMATE_STATUS.IDLE;
            this._interactionHandler && InteractionManager.clearInteractionHandle(this._interactionHandler);
        });
    }

    _hide(updateState) {
        if (!this._layouted) {
            // 如果还没有 layout，则暂存“隐藏”操作，等 layout 完成后执行“隐藏”操作
            this._showActionStashed = false;
            this._hideActionStashed = true;
            return;
        }
        if (this._animateStatus === ANIMATE_STATUS.HIDDING) {
            // 防止重复动画
            return;
        }
        this._showAnimation && this._showAnimation.stop();
        this._animateStatus = ANIMATE_STATUS.HIDDING;
        // 注册动画，方便调度
        this._interactionHandler && InteractionManager.clearInteractionHandle(this._interactionHandler);
        this._interactionHandler = InteractionManager.createInteractionHandle();
        Animated.timing(
            this._backdropAnimatedValue,
            {
                toValue: 0,
                duration: 300,
                useNativeDriver: true,
            }
        ).start();
        this._hideAnimation = Animated.timing(
            this.state.popupAnimatedValue,
            {
                toValue: 0,
                duration: 300,
                useNativeDriver: Platform.OS !== 'ios' ? true : false,
            }
        );
        this._hideAnimation.start(() => {
            // 动画结束
            if (this._animateStatus !== ANIMATE_STATUS.SHOWING) {
                updateState();
                this._layouted = false;
            }
            this._animateStatus = ANIMATE_STATUS.IDLE;
            requestAnimationFrame(() => {
                this._interactionHandler && InteractionManager.clearInteractionHandle(this._interactionHandler);
            });
        });
    }

    _dismiss() {
        const { onClose } = this.props;
        onClose && onClose();
    }

    _handleLayout(e) {
        const isContentHeightChanged = this._contentHeight !== e.nativeEvent.layout.height;
        this._contentHeight = e.nativeEvent.layout.height;
        if (this._layouted && !isContentHeightChanged) { return; }
        if (isContentHeightChanged && this.state.visible) {
            this.setState({
                contentHeight: this._contentHeight,
            });
            if (this._onceLayouted) {
                // 如果高度变化、弹出框为显示状态且已经 layout 过一次，则调整弹出框位置
                this._adjustPosition();
            }
        }
        this._layouted = true;
        this._onceLayouted = true;
        // 消费暂存的操作
        if (this._showActionStashed) {
            this._show();
            this._showActionStashed = false;
        }
        if (this._hideActionStashed) {
            this._hide(() => {
                this.setState({
                    visible: false,
                });
            });
            this._showActionStashed = true;
        }
    }

    _handleRequestClose() {
        this.props.onClose && this.props.onClose();
    }

    render() {
        const { children, homeIndicatorStyle } = this.props;

        return (
            <Modal transparent={true} visible={this.state.visible} onRequestClose={this._handleRequestClose}>
                {/* popover 的背景，点击后关闭 popover */}
                <Animated.View
                    style={[
                        styles.backdrop,
                        { opacity: this._backdropAnimatedValue }
                    ]}
                >
                    <TouchableOpacity
                        style={styles.backdropWrapper}
                        activeOpacity={1}
                        onPress={() => this._dismiss()}
                    />
                </Animated.View>
                <Animated.View
                    style={[
                        styles.container,
                        {bottom: -1 * this.state.contentHeight},
                        {transform: [{translateY: this.state.popupAnimatedValue}]},
                    ]}
                    onLayout={this._handleLayout}
                >
                    {children}
                    {/* iPhoneX 下巴占位 */}
                    {isIPhoneX() ? <View style={[styles.homeIndicator, homeIndicatorStyle]} /> : null}
                </Animated.View>
            </Modal>
        );
    }
}

Popup.propTypes = {
    visible: PropTypes.bool,
    onClose: PropTypes.func,
    homeIndicatorStyle: View.propTypes.style,
};

Popup.defaultProps = {
    visible: false,
};

const styles = StyleSheet.create({
    container: {
        position: 'absolute',
        bottom: -WINDOW_HEIGHT - 100,
        left: 0,
        right: 0,
    },
    backdrop: {
        flex: 1,
        backgroundColor: theme('color_mask'),
    },
    backdropWrapper: {
        flex: 1,
    },
    homeIndicator: {
        height: 34,
        backgroundColor: theme('color_white'),
    },
});

export default Popup;
