// @souche-ui/srn-ui/src/components/Tabs/styles/index.style.js
import { StyleSheet, Dimensions } from 'react-native';
import FontSize from '../../../const/FontSize';
import theme from '../../../theme';
const styles = {
    container: {
        flex: 1,
    },
    pageContainer: {
        flex: 1,
        width: Dimensions.get('window').width
    },
    scrollTabsContent: {
        // flex:1,
        flexDirection: 'row',
    },
    content: {
        height: 44,
        backgroundColor: theme('color_white'),
        borderBottomWidth: StyleSheet.hairlineWidth,
        borderBottomColor: theme('color_line_b'),
    },
    tabs: {
        flex:1,
        flexDirection: 'row',
        justifyContent: 'space-around',
        alignItems: 'center',
    },
    tabsUnderline: {
        height:2,
        backgroundColor: theme('color_primary'),
    },
    scrollTabs: {
        backgroundColor: theme('color_white'),
    },
    tab: {
        flex:1,
        flexDirection: 'row',
        alignItems: 'center',
        justifyContent:'center',
        paddingHorizontal: 12,
    },
    tabText: {
        color: theme('color_text_title'),
        fontSize: FontSize.P2,
    },
};


export default styles;
